README.md
# 🛫 RDGliderViewController-Swift 🛬

[![Build Status](https://travis-ci.org/gelemias/RDGliderViewController-Swift.svg?branch=master)](https://travis-ci.org/gelemias/RDGliderViewController-Swift) [![Coverage Status](https://coveralls.io/repos/github/gelemias/RDGliderViewController-Swift/badge.svg?branch=master)](https://coveralls.io/github/gelemias/RDGliderViewController-Swift?branch=master) [![License: MIT](https://img.shields.io/github/license/mashape/apistatus.svg)](https://opensource.org/licenses/MIT) [![Carthage compatible](https://img.shields.io/badge/Carthage-compatible-4BC51D.svg?style=flat)](https://github.com/Carthage/Carthage)


Control in *Swift* for a draggable ViewController gliding over another ViewController, if you are looking for the equivalent in *Objective-C* go [here](https://github.com/gelemias/RDGliderViewController/).

RDGliderViewController is a view controller that manages a scrollable view using one side as a sliding point, choosing between Left-to-Right, Top-to-Bottom, Right-to-Left and Bottom-to-Top, and setting an array of offsets as percent values to determine the gliding view steps, as described bellow.

<img src="./img/1.gif" width="180"> <img src="./img/2.gif" width="180"> <img src="./img/3.gif" width="180"> <img src="./img/4.gif" width="180">

## Example

A good way to learn how to use RDGliderViewController is to go through the example app. Take a look at the tests as well for a more thorough usage.

To run the example project, clone the repo, and open `./Example/GliderSample.xcodeproj`

## Installation

### Using CocoaPods
RDGliderViewController is available through [CocoaPods](http://cocoapods.org), a dependency manager for Swift and Objective-C Cocoa projects. To install
it, simply add the following line to your [Podfile](https://guides.cocoapods.org/using/getting-started.html):

```ruby
pod 'RDGliderViewController-Swift'
```

### Using Carthage

[Carthage](https://github.com/Carthage/Carthage) is a decentralized dependency manager that builds your dependencies and provides you with binary frameworks.

You can install Carthage with [Homebrew](http://brew.sh/) and then integrate **RDGliderViewController-Swift** into your Xcode project by adding the following line to your [Cartfile](https://github.com/Carthage/Carthage/blob/master/Documentation/Artifacts.md#cartfile):

```ruby
github "gelemias/RDGliderViewController-Swift"
```

### Manual Installation

For an old fashion installation you can directly add the **header** and **implementation** files for `RDGliderViewController` and `RDScrollView` to your project.

- Download the latest code version or add the repository as a git submodule to your git-tracked project.

- Drag and drop the folder `RDGliderViewController/` onto your project and make sure to select **Copy items if needed**.

- Add `#import RDGliderViewController` at the beginning of your class.

## Usage

RDGliderViewController is a very simple to use controller that mainly depends on three values to work on, *Orientation type* which is determines the scrolling side, the *List of Offsets* which represent the steps of expansion and contraction and the content itself with is a view controller completely independent.

### Creation

Use a the following custom init to instantiate RDGliderViewController:

`public init(parent: UIViewController, WithContent content: RDGliderContentViewController, AndType type: RDScrollViewOrientationType, WithOffsets offsets: [NSNumber])
`

e.g.
```Swift
let content:RDGliderContentViewController = RDGliderContentViewController.init()

self.rightToLeftGlideVC = RDGliderViewController.init(parent: self,
                                                      WithContent: content,
                                                      AndType: .RDScrollViewOrientationRightToLeft,
                                                      WithOffsets: [0, 0.6, 1])
```

And that's all you need to do, if you run your app, an instance of UIViewController should be scrollable from Right to Left on top of `self` view controller.

### Content and Container

The content view controller should inherit from RDGliderContentViewController to inherit the properly resize when rotating, is treated as a child view controller of this one.


```Swift

class ContentViewController: RDGliderContentViewController {

// Your content class Here

}

```

Mentioned that not much more is required, only keep in mind that if you do not define a size for the view, RDGliderViewController will it resize it to match container's.

Content view controller can be afterwards as well but always along offsets and orientation since these three properties are dependent of each other:

`public func setContentViewController(Content content: RDGliderContentViewController,
                                         AndType type: RDScrollViewOrientationType,
                                  WithOffsets offsets: [NSNumber])`

### Orientation types

`RDScrollViewOrientationType` represent the four sides of the display plus the sliding direction.

```Swift
public enum RDScrollViewOrientationType: Int {

    case RDScrollViewOrientationUnknown

    case RDScrollViewOrientationLeftToRight

    case RDScrollViewOrientationBottomToTop

    case RDScrollViewOrientationRightToLeft

    case RDScrollViewOrientationTopToBottom
}
```

if content view controller defines a fixed size, this size will be respected during rotation, for horizontal scrolls (Left-to-Right and Right-to-Left) will resize *Width* and for Vertical (Bottom-to-Top and Top-to-Bottom) will reize *Height*


### Offsets and OffsetIndex

this `offsets: [NSNumber]` property is a list of steps used when either expanding / collapsing or dragging the list is represented in percent % related with the length of the content view controller, e.g. given the list [0, 0.5, 1] for a lenght *400px* would be this offset steps of [0px, 200px, 400px]

`offsets[OffsetIndex].floatValue * contentViewController.lenght`

For inverted OrientationTypes (RDScrollViewOrientationLeftToRight and RDScrollViewOrientationTopToBottom), the offset list are also inverted so for example for a list [0.2, 0.4, 0.6, 1] will be returned [0.8, 0.6, 0.4, 0]

### Expand, Collapse and Close

By default RDGliderViewController is draggable but when setting an offset of `0.0f%` it will be outside of the screen on starting point, to Increase the position of the view by one in the list of offsets use:

`public func expand()`

to Decrease the position of RDGliderViewController by one in the list of offsets:

`public func collapse()`

and *close* which moves the View directly to the first offset which is the default position:

`public func close()`

All of these methods simple change the offsetIndex when calling this method.

`public func changeOffset(to offsetIndex: Int, animated: Bool)`

## License

RDGliderViewController is released under the MIT license. See [LICENSE](https://github.com/gelemias/RDGliderViewController/blob/develop/LICENSE) for details.


## Change log

See [CHANGELOG](https://github.com/gelemias/RDGliderViewController/blob/develop/CHANGELOG.md) for details.


CHANGELOG.md
## CHANGELOG

#### Latest version -  1.0.23
+ 2017-11-22 *[FIX] Fixed issue happening with changing margin and starting from different offset than 0*

#### Previously released:
+ 2017-05-30 *[ENV] bump version*
+ 2017-05-30 *[FIX] fixed access control to suit swift 3*
+ 2017-05-30 *[FIX] fixed access control*
+ 2017-05-30 *[Env] update build version to 1.0.16*
+ 2017-05-30 *[ENV] update build version - to upload to cocoapods*
+ 2017-05-30 *[REF] ignored length warnings*
+ 2017-05-30 *[REF] refactor more using swiftlint*
+ 2017-05-29 *[REF] fixed code according to SwiftLint*
+ 2017-04-28 *[TEST] Added tests in ObjC to reused from the other repo*
+ 2017-04-27 *[TEST] added Objc tests -needs to be fixed*
+ 2017-04-27 *[ENV] updated README file*
+ 2017-04-27 *[ENV] fixed CI build*
+ 2017-04-27 *[FIX] fixed several issues*
+ 2017-04-27 *[DEV] implemented constraints for reversed glider view controllers*
+ 2017-04-27 *[DEV] completed first implementation*

#### Previously released:


Example/GliderSample/ViewController.swift
//
//  ViewController.swift
//  GliderSample
//
//  Created by Guillermo Delgado on 25/04/2017.
//  Copyright © 2017 Guillermo Delgado. All rights reserved.
//

import UIKit

extension ViewController: RDGliderViewControllerDelegate {
    func glideViewControllerWillExpand(glideViewController: RDGliderViewController) {
        print("Will Expand")
    }

    func glideViewControllerDidExpand(glideViewController: RDGliderViewController) {
        print("Did Expand")
    }

    func glideViewControllerWillCollapse(glideViewController: RDGliderViewController) {
        print("Will Collapse")
    }

    func glideViewControllerDidCollapse(glideViewController: RDGliderViewController) {
        print("Did Collapse")
    }

    func glideViewController(glideViewController: RDGliderViewController, hasChangedOffsetOfContent offset: CGPoint) {
        print("New Offset: " + NSStringFromCGPoint(offset))
    }
}

class ViewController: UIViewController {

    var rightToLeftGlideVC: RDGliderViewController?
    var bottomToTopGlideVC: RDGliderViewController?
    var leftToRightGlideVC: RDGliderViewController?
    var topToBottomGlideVC: RDGliderViewController?

    override func viewDidLoad() {
        super.viewDidLoad()
        initRightToLeftGlideView()
        initBottomToTopGlideView()
        initLeftToRightGlideView()
        initTopToBottomGlideView()
    }

    func initRightToLeftGlideView() {
        let content: RDGliderContentViewController = RDGliderContentViewController.init()
        content.view.backgroundColor = UIColor.red
        content.showShadow = true
        content.cornerRadius = 10.0

        self.rightToLeftGlideVC = RDGliderViewController.init(parent: self,
                                                              WithContent: content,
                                                              AndType: .RDScrollViewOrientationRightToLeft,
                                                              WithOffsets: [0, 0.6, 1])
        self.rightToLeftGlideVC?.marginOffset = 0.0
        self.rightToLeftGlideVC?.delegate = self
    }

    func initBottomToTopGlideView() {
        let content: RDGliderContentViewController = RDGliderContentViewController.init(length: 200.0)
        content.view.backgroundColor = UIColor.purple
        content.showShadow = true
        content.cornerRadius = 10.0

        self.bottomToTopGlideVC = RDGliderViewController.init(parent: self,
                                                              WithContent: content,
                                                              AndType: .RDScrollViewOrientationBottomToTop,
                                                              WithOffsets: [0.6, 0, 0.3, 1])
        self.bottomToTopGlideVC?.delegate = self
    }

    func initLeftToRightGlideView() {
        let content: RDGliderContentViewController = RDGliderContentViewController.init()
        content.view.backgroundColor = UIColor.orange
        content.showShadow = true
        content.cornerRadius = 10.0

        self.leftToRightGlideVC = RDGliderViewController.init(parent: self,
                                                              WithContent: content,
                                                              AndType: .RDScrollViewOrientationLeftToRight,
                                                              WithOffsets: [0, 0.2, 0.4, 0.6, 0.8, 1])
        self.leftToRightGlideVC?.delegate = self
    }

    func initTopToBottomGlideView() {
        let content: RDGliderContentViewController = RDGliderContentViewController.init()
        content.view.backgroundColor = UIColor.magenta
        content.showShadow = true
        content.cornerRadius = 10.0

        self.topToBottomGlideVC = RDGliderViewController.init(parent: self,
                                                              WithContent: content,
                                                              AndType: .RDScrollViewOrientationTopToBottom,
                                                              WithOffsets: [0, 0.5, 1])
        self.topToBottomGlideVC?.delegate = self
    }

// MARK: - Actions

    @IBAction func btnPressed(_ sender: UIButton) {

        var gliderVC: RDGliderViewController?

        switch sender.tag {
        case 1:
            gliderVC = rightToLeftGlideVC
        case 2:
            gliderVC = bottomToTopGlideVC
        case 3:
            gliderVC = leftToRightGlideVC
        default:
            gliderVC = topToBottomGlideVC
        }

        if gliderVC != nil {
            if gliderVC!.currentOffsetIndex < Int(gliderVC!.offsets.count) - 1 {
                gliderVC?.expand()
            } else {
                gliderVC?.shake()
            }
        }
    }
}
